pandas/core/algorithms.py

```diff
@@ -1829,11 +1830,25 @@ def diff(arr, n: int, axis: int = 0):
     -------
     shifted
     """
+    from pandas.core.arrays import PandasDtype
 
     n = int(n)
     na = np.nan
     dtype = arr.dtype
 
+    if dtype.kind == "b":
+        op = operator.xor
+    else:
+        op = operator.sub
+
+    if isinstance(dtype, PandasDtype):
+        # PandasArray cannot necessarily hold shifted versions of itself.
+        arr = np.asarray(arr)
+        dtype = arr.dtype
+
+    if is_extension_array_dtype(dtype):
```

Comment:
should you not be able to remove the timedelta / datetime stuff from below?

Reply:
Timedelta still goes through the same code path as master since `TimedeltaArray.dtype` isn't an extension dtype, it's a NumPy dtype. And you can also have 2D datetime / timedelta arrays passed here.

Reply:
OK, I think this works. Added a test for categorical grouper, so I think both cases are covered.